nanoRMS2: Documentation
=======================


What is nanoRMS2?
-----------------
nanoRMS2 is a pipeline for detection of DNA/RNA modifications. 
The pipeline consists of three steps / modules:

- get_feautres: retrieve & store features in BAM file
- predict_mods: train ML models for modified bases,
  estimate per-position stoichiometry and much more
- encode_mods: predict & encode per-read modification status
  in BAM file using ML models generated by predict step. 

All these scripts can be run separately or as a pipeline by executing `nanoRMS2/run`.  
You can find more the details in the
Supplementary Methods in nanoRMS2 paper_.

.. image:: pipeline.png
   :align: center
   :width: 75%


What do I need to run nanoRMS2?
-------------------------------
To run modPhred, you will need:

0. :doc:`all dependencies installed <install>`
1. reference sequence (FastA)
2. raw ONT data (Fast5) for two samples that differ in modification status
   
   - without modifications (PCR, IVT, KO) or with lower modification stoichiometry (KD, stress conditions)
   - with at least some reads being modified (WT)

nanoRMS2 works best if modifications are deposited in:

- multiple genome/transcriptome positions
- sequence motif (at least weakly defined, but it can work even if the motif isn't present). 

:doc:`More information about running the pipeline <usage>`.


Getting help
------------
If you have a question/issue after familiarising with this documentation, 
please have a look at `issues <https://github.com/novoalab/nanoRMS2/issues?q=>`_
users reported so far. If you don't find the solution, please open a new issue.

.. _paper:

Citation
--------
If you find this work useful, please cite:


Please, consider citing :doc:`dependencies <install>` as well.


.. toctree::
   :maxdepth: 3
   :hidden:
   :caption: First steps
	     
   install
   test
   usage
   output
   igv

.. toctree::
   :maxdepth: 3
   :hidden:
   :caption: Methods
	     
   training
   
